### PR TITLE
Replace timer dueMap with rbtree to avoid apocalypse

### LINKED ISSUE #2615: Fix time apocalypse in timer.ts
In js/timers.ts, next TODO is told.
```typescript
// We'll subtract EPOCH every time we retrieve the time with Date.now(). This
// ensures that absolute time values stay below UINT32_MAX - 2, which is the
// maximum object key that EcmaScript considers "numerical". After running for
// about a month, this is no longer true, and Deno explodes.
// TODO(piscisaureus): fix that ^.
```

So fix this. 😃

### PR DESCRIPTION
**Pull request by kevinkassimo**
Fixes #2615 

This is nothing more than a line-by-line replacement of dueMap with dueTree, a red-black tree. No timer implementation logic is changed.

The red-black tree implementation is copied from https://github.com/vadimg/js_bintrees/blob/master/lib/rbtree.js , with types annotated.

### CODE REVIEW

**Comment by kevinkassimo** [comment]
Hmm CI is failing for some unrelated test...

**Comment by kevinkassimo** [comment]
@ry This change ensures that the iteration would always be inorder even under large numbers.
Original TODO:
https://github.com/denoland/deno/blob/65d9286203cf239f68c6015818e82e8521e600a1/cli/js/timers.ts#L18-L22
I am not so sure about the details of implementing using tokio timers, since currently we implement using a shared global tokio timer. Are we planning to split into multiple tokio delays and poll each future separately?

**Comment by ry** [comment]
> Are we planning to split into multiple tokio delays and poll each future separately?

Yes - it's simpler for our implementation. Although it's potentially less optimal - however I very much doubt we're anywhere near having it be a bottleneck for anyone - and it's not measured. So I'd prefer to do the simpler thing: mapping JS timers to tokio timers in a one-to-one fashion. (Note: this was my original implementation see https://github.com/denoland/deno/blob/4fd2b19f640d19e57511eb142b63e16c879ef6fd/js/timers.ts and https://github.com/denoland/deno/blob/4fd2b19f640d19e57511eb142b63e16c879ef6fd/src/handlers.rs#L371-L393)

**Comment by piscisaureus** [comment]
I think landing this is still a net improvement over the status quo.

**Comment by ry** [comment]
How much size does it add to the executable?

**Comment by kevinkassimo** [comment]
@ry ~0.09M

File: cli/js/timers.ts [hunk 1]
```diff
@@ -15,13 +16,8 @@ interface Timer {
   scheduled: boolean;
 }
 
-// We'll subtract EPOCH every time we retrieve the time with Date.now(). This
-// ensures that absolute time values stay below UINT32_MAX - 2, which is the
-// maximum object key that EcmaScript considers "numerical". After running for
-// about a month, this is no longer true, and Deno explodes.
-// TODO(piscisaureus): fix that ^.
+// We'll subtract EPOCH every time we retrieve the time with Date.now().
```
Comment (piscisaureus) [inline cli/js/timers.ts]: I think that with this patch you could remove EPOCH altogether.

Comment (kevinkassimo) [reply cli/js/timers.ts]: Yeah makes sense

**Pull request merged**